src/main/java/bio/terra/app/controller/RepositoryApiController.java

```diff
@@ -361,6 +362,21 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
             "User is not authorized to create snapshots for these datasets " + unauthorized);
     }
 
+    @Override
+    public ResponseEntity<JobModel> createSnapshotWithProvidedIds(
+        @Valid @RequestBody SnapshotProvidedIdsRequestModel spiRequestModel) {
+        AuthenticatedUserRequest userReq = getAuthenticatedInfo();
+        Snapshot snapshot = snapshotService.makeSnapshotFromSnapshotProvidedIdsRequest(spiRequestModel);
+        List<SnapshotSource> unauthorized = getUnauthorizedSources(snapshot, userReq);
+        if (unauthorized.isEmpty()) {
```

Comment:
Nit: I'd prefer reversed logic, like
```suggestion
        if (!unauthorized.isEmpty()) {
            throw new IamUnauthorizedException("User is not authorized to create snapshots for these datasets " + unauthorized);
         }
        String jobId = snapshotService.createSnapshotWithProvidedIds(spiRequestModel, userReq);
         // we can retrieve the job we just created
        return jobToResponse(jobService.retrieveJob(jobId, userReq));         
```
but that is just my style preference.